04875 bucket restructure hip (#1)
mgoelswirlds/hedera-improvement-proposal: HIP/hip-641.md

---

> hip: 641

Author: I am not sure how the number is picked. But the doc said it is up to the author.

title: Bucket restructuring for changing stream data folder structure
author: Mugdha Goel (@mgoelswirlds)

> working-group: Richard Bair (@rbair23), Nick Poorman (@nickpoorman), Jasper Potts, Dan aobizyu

Reviewer: Add me too

type: Standards Track

> category: Mirror

Reviewer: ```suggestion
category: Mirror, Service
```

needs-council-approval: Yes
status: Draft
last-call-date-time: 
created: 2023-01-09
discussions-to: 
requires: 
replaces: 
updated: 
---

## Abstract

> We propose changing the current stream data bucket folder structure from node account ID being shard and node ID based in the path.

Reviewer: ```suggestion
We propose changing the current stream data bucket folder structure from node account ID based to node ID based. Additionally, we propose incorporating shard into the path to be future compatible with sharding when it's enabled.
```

Current Structure:
```
    accountBalances/balance0.0.3/
    eventsStreams/events_0.0.3/

>     recordstreams/record0.0.3/

Reviewer: We could also mention how this format is very inconsistent in regards to casing, separators, and duplicating stream name.

Author: I need to understand this better.

```
    
Proposed Structure:
```
    {shard}/{nodeID}/balance/
    {shard}/{nodeID}/event/
    {shard}/{nodeID}/record/
```

## Motivation

> The current folder structure uses an account ID per node. Account IDs are not ideal as a council member will want to change their account ID over time. To future-proof this folder

Reviewer: Somewhere we should define the difference between a node's account ID and the node ID. How are the two assigned and change over their lifetime.

structure we will want to move from account IDs to shard and node ID in the path.

## Rationale

> By adapting the new node ID based path we provide the council members the flexibility to change their account ids as well as allow only node specific access to the buckets for future community nodes inclusion.

Reviewer: Doesn't talk about why the need for rollups

## User stories

- As a consumer of the records and balances stream data bucket, the existing Mirror nodes need to be able to be able to read files from the new bucket path as well as keep supporting the old path till it is obsolete.
- The consensus nodes need to start writing the stream data to the new bucket location.
- A new lambda would be created which aggregates/rolls up data from the current bucket path into larger compressed files, to facilitate new mirror nodes to ingest data at a higher speed.
  
## Specification

This HIP process to change the stream data bucket folder structure and in the process roll up data for faster ingestion.

It is thus proposed to be a phased approach with three phases.

#### Phase 1:
Update mirror node downloader to read from the new `{shard}/{nodeID}/*` nodeId based path as well as support reading from `*/*0.0.3/` account id based path.

##### Algorithm:

>  - Add a pathType property with values AUTO, ACCOUNT_ID, and NODE_ID with a default of AUTO

Reviewer: Quote the enum values and other properties here and elsewhere.

```suggestion
 - Add a `pathType` property with values `AUTO`, `ACCOUNT_ID`, and `NODE_ID` with a default of AUTO
```

>  - Add a pathRefreshInterval property with value for eg:1m

Reviewer: ```suggestion
 - Add a `pathRefreshInterval` property with a default value of `1m`
```

    If pathType is AUTO:
    - Attempt to list signatures from a path with node IDs
    - If successful, update a per node timestamp and boolean to ensure node ID is used until pathRefreshInterval elapses
    - If no results, attempt to list signatures from a path with node account IDs

>     - If successful, update a per node timestamp and boolean to ensure account ID is used until pathRefreshInterval elapses

Reviewer: Why did we reverse the logic from ticket? I think it would be more efficient to query node account IDs first since once it transitions to node ID it can remain that way indefinitely. I've updated the [ticket](https://github.com/hashgraph/hedera-mirror-node/issues/4388) with some fine tuning of the algorithm that can be used here.

Author: Ok. I reversed it , thinking node account ID path is a backup , so if we didn't find the file at the first location, we would do this. But your reasoning makes sense

   
In the existing scenario the default pathType would be ACCOUNT_ID, which would then be changed to AUTO when we are close to approaching Phase 2.

> After a significant amount of time has passed after launching Phase 2, we will remove these properties to default the downloading from the NODE_ID based path.

Reviewer: ```suggestion
After a significant amount of time has passed after launching Phase 2, we will remove these properties to default the downloading from the node ID based path.
```

#### Phase 2:
Update the consensus nodes to write stream files to the new `{shard}/{nodeID}/*` nodeId based path.

#### Phase 3:
Create a new lambda function which performs data rollup along with compression.

##### Algorithm:

> - Read stream files from current location 

Reviewer: Read how many stream files? The algorithm doesn't discuss that lambda would have to keep track of where it left off and that later it would delete stream files.

Author: I still do not know how many stream files, since that will be determined by performance testing. But yes the other parts of that question will need to be addressed.

    ``` 
        accountBalances/balance0.0.3/ 
        eventsStreams/events_0.0.3/     
        recordstreams/record0.0.3/ 
    ```

> - Calculate the node Id based on the current node account id. NodeId = (number portion of) nodeAccountId - 3

Reviewer: Algorithm doesn't discuss that lambda would hash each file so it can discard duplicates.

  ``` Node for balance0.0.4 would be 4-3 = 1  ```
  The new path would be:
  ``` 
      {shard}/record/
      {shard}/balance/

>       {shard}/event/

Author: This is for the aggregated files.It wouldn't have nodeId right?

>       {shard}/event/

Reviewer: True, missed originally that it was for aggregated.

  ```
- The new compressed file structure would be as follows:
    Example:
    ```

>             2023-01-11T23_21.tar.zst

Reviewer: There's no need to recompress compressed files.
```suggestion
            2023-01-11T23_21.tar
```

Reviewer: I said this previously, but I think your testing actually showed there was improvement with both inner and outer compression, right?

Author: That is correct.

>             2023-01-11T23_21.tar.zst

Reviewer: We should talk about why zstd over gzip since that is a change. We should discuss with ops/services if that's something they can change on the consensus node since it's quite inefficient to recompress files in the lambda if we can help it. We'd have to have some good evidence to convince about zstd's benefits.

             - 0(This is the node id)
               - 2023-01-11T23_21_52.007485807Z.rcd.zst
               - 2023-01-11T23_21_50.006107818Z.rcd.zst
               - 2023-01-11T23_21_46.003350460Z.rcd.zst

>                - 2023-01-11T23_21_44.001332036Z.rcd.zst

Reviewer: Where are the signatures? Why would the RCD be in a node ID based path in the file?

Author: I forgot to add the signature file. Will do that.I misremembered your earlier statement about having signatures in node id based path.Will correct this.

Author: I still don't see a problem having the rcd and signature files together in the node based path. Unless there is something I am missing here.

created: 2023-01-09
discussions-to: 
requires: 
replaces: 
updated: 
---

## Abstract

We propose changing the current stream data bucket folder structure from node account ID based to node ID based. Additionally, we propose incorporating shard into the path to be future compatible with sharding when it's enabled.

## Motivation

Node account IDs are Hedera account IDs, which each node has obtained. These can be changed over time.
In the future we plan to introduce community nodes, thus increasing the number of unique nodes across the Hedera network. Granting specific node ID based access to buckets will be of utmost importance for security reasons.
The current folder structure uses an account ID per node. Account IDs are not ideal as a council member will want to change their account ID over time. To future-proof this, we will want to move from account IDs to shard and node ID in the path.

## Rationale

- There have been council members that want to change their account id but haven’t been able to. Aliases can be assigned only on create. So they might want to create a new account id with an alias.
- There might also be a need to recreate an account for security reasons.
 

>  By adapting the new node ID based path we provide the council members the flexibility to change their account IDs as well as allow only node specific access to the buckets for future community nodes inclusion.

Reviewer: Formatting here is inconsistent. Should we make this line another list element or remove all list elements?

Author: The reason I did this was the statement not as a list element was relevant to the above two. But will reformat it.

## Specification

This HIP is to change the stream data bucket folder structure and in the process roll up data for faster ingestion.

Current Structure:
```
    accountBalances/balance0.0.3/
    eventsStreams/events_0.0.3/
    recordstreams/record0.0.3/
```
    
Proposed Structure:
```
    {shard}/{nodeID}/balance/YYYY/MM/DD
    {shard}/{nodeID}/event/YYYY/MM/DD
    {shard}/{nodeID}/record/YYYY/MM/DD
```
##### Problems with the current structure naming:
- Stream name is duplicated in `eventsStreams` and `recordstreams`. We are missing stream from `accountBalances`.
- Casing is inconsistent in `eventsStreams` and `recordstreams`.
- After the separator we are repeating stream names. `events_0.0.3` has an underscore where as the others don't.

The new structure proposes consistent naming.In the future when we start supporting multiple shards, the new bucket structure will automatically work for it without any required changes.

It is thus proposed to be a phased approach with three phases.

#### Phase One:
Update mirror node downloader to read from the new `{shard}/{nodeID}/*` nodeId based path as well as support reading from `*/*0.0.3/` account id based path.

##### Algorithm:

>  - Add a `pathType` property with values `AUTO`, `ACCOUNT_ID`, and `NODE_ID` with a default of `AUTO`

Reviewer: Mentions defaulting to AUTO twice. Though we can also just omit the line `Later once consensus nodes get closer to writing to the new path adjust default to AUTO` and always default to AUTO since that's an implementation/cost saving detail.

```suggestion
 - Add a `pathType` property with values `AUTO`, `ACCOUNT_ID`, and `NODE_ID` with a default of `ACCOUNT_ID`
```

Reviewer: Yes, I wrote the algorithm and agree with it for implementation on the mirror node. I'm saying it's irrelevant to the HIP and just a cost saving mechanism that doesn't need to be called out. For example, if we're saying end of march go live then it wouldn't even make sense to ever have a period that's not AUTO cause it introduces risk of people staying on ACCOUNT_ID.

 - Later once consensus nodes get closer to writing to the new path adjust default to `AUTO`
 - Add a `pathRefreshInterval` property with a default value of `1m`
    If pathType is `AUTO`:
    - Attempt to list signatures from  a path with node account IDs
    - If successful, update a per node timestamp and boolean to ensure account ID is used until pathRefreshInterval elapses
    - If no results, attempt to list signatures from a path with node IDs
    - If successful, set `pathType` as `NODE_ID` so the node ID path is used going forward
   
In the existing scenario the default pathType would be `ACCOUNT_ID`, which would then be changed to AUTO when we are close to approaching Phase two.
After a significant amount of time has passed after launching Phase two, we will remove these properties to default the downloading from the node ID based path.

#### Phase Two:
Update the consensus nodes to write stream files to the new `{shard}/{nodeID}/*` nodeId based path.

#### Phase Three:

> Create a new lambda function which performs data rollup along with compression.

Reviewer: We should call out that another reason for doing the rollup is to avoid having to rename millions of files. We should also mention this in the rejected ideas. Renaming millions of files has to be done one by one via S3 APIs and would take months. With rollup we can create separate aggregated files while leaving individual files intact for longer period until we've verified authenticity of rollup.

Author: I do not understand why this would be a rejected idea?

##### Compression zstd vs gz:
- zstd has a better compression ration that gzip.
- (zstd:gzip)
  - Faster compression - About a 5:1  ratio.
  - Faster decompression - About a 3:1 ratio.

##### Algorithm:
- Read stream files from current location 
    ``` 
        accountBalances/balance0.0.3/ 
        eventsStreams/events_0.0.3/     
        recordstreams/record0.0.3/ 
    ```
    
- The new path for aggregated files would be:
    
    ``` 
        {shard}/record/
        {shard}/balance/
        {shard}/event/
    ```
- The lambda will maintain a track of the read files.(To be designed)
- It will decompress the file, if it is compressed.
- Calculate a hash of the file.

> - Maintain an in-memory DB which has all the hashes of aggregated files.

Reviewer: DB is probably the wrong terminology. It should probably just be a map. Also, it wouldn't be all the hashes, just the hashes for the current filename being processed.

Author: You mean the current aggregated file correct?

Reviewer: Deduplication wouldn't need a map of all files within the rollup file. Just the current RCD filename being processed in lexographical order.

- Lookup to see if the hash already exists in the DB to be aggregated (to avoid duplicates).
- Compress the file using zst if it not a signature or an event stream file.
- Monitor the size/time of the potential aggregated file.
- For aggregation:
  - Calculate the node Id based on the current node account id. NodeId = (number portion of) nodeAccountId - 3
  ``` Node for balance0.0.4 would be 4-3 = 1  ```
  - The new compressed file structure would be as follows:
    Example:
    ```
            2023-01-11T23_21.tar.zst
               - 2023-01-11T23_21_52.007485807Z.rcd.zst
               - 2023-01-11T23_21_50.006107818Z.rcd.zst
               - 2023-01-11T23_21_46.003350460Z.rcd.zst
               - 2023-01-11T23_21_44.001332036Z.rcd.zst
               - 0(This is the node id)
                 - 2023-01-11T23_21_52.007485807Z.rcd_sig
                 - 2023-01-11T23_21_50.006107818Z.rcd_sig
                 - 2023-01-11T23_21_46.003350460Z.rcd_sig
                 - 2023-01-11T23_21_44.001332036Z.rcd_sig
     ```
     The aggregation in the new files would be size/time based, as will be determined by performance testing.

> - Once the limit has reached aggregate the file and write it to the new bucket location. 

Reviewer 2: ```suggestion
- Once the limit has been reached, aggregate the file and write it to the new bucket location. 
```

- Delete the existing files that have been read from the current bucket location.

## Backwards Compatibility

This HIP proposes breaking changes. All existing mirror node operators would need to change code to support reading from the new as well as the old bucket location. According to the current plan for phase one and two, operators will have ample time until the completion of phase two and a significant amount of notification period, before the mirror nodes will become backwards incompatible.

## Mirror Node Implications
The existing mirror nodes would need to implement phase one.

## Dev ops Implications
After the completion of phase one and having being notified, dev ops will need to implement phase two.

## Security Implications

With this change, we are improving security. Previously, each node had to be granted to three separate paths since we had stream types as the top level folders.

> Write access to the new node id based path will be granting access to only specific node path buckets. This would prevent a malicious node to have write access or corrupt other stream files

Reviewer 2: ```suggestion
Write access to the new node id based path will be granting access to only specific node path buckets. This would prevent a malicious node to have write access or corrupt other stream files.
```

The rollup process could be seen as centralized and a single point of failure. Since, the lambda will be deleting files it has already aggregated, this might be pursued as a problem.

> All files are cryptographically signed by each node. Same data will be in each file. We would be just downloading data and aggregating it. Record file bytes would be unchanged. Signatures would still be present. We could use it to verify the authenticity as it is currently being done in the mirror node.

Reviewer: Wondering what sort of ongoing verification we should do before deletion. After we create this large file how do we verify it's not corrupt, has all the files with no break in running hash or missing signatures, etc?

Author: Need to dig in a little more for this one.

type: Standards Track
category: Mirror, Service
needs-council-approval: Yes
status: Draft
last-call-date-time: 
created: 2023-01-09
discussions-to: 
requires: 
replaces: 
updated: 
---

## Abstract

We propose changing the current stream data bucket folder structure from node account ID based to node ID based. Additionally, we propose incorporating shard into the path to be future compatible with sharding when it's enabled. We are also proposing a rollup mechanism to reduce cloud storage duplication and improve historical sync times of mirror nodes.

## Motivation

Node account IDs are Hedera account IDs, which each node has obtained. These can be changed over time.
In the future we plan to introduce community nodes, thus increasing the number of unique nodes across the Hedera network. Granting specific node ID based access to buckets will be of utmost importance for security reasons.
The current folder structure uses an account ID per node. Account IDs are not ideal as a council member will want to change their account ID over time. To future-proof this, we will want to move from account IDs to shard and node ID in the path.

## Rationale

> - There have been council members that want to change their account id but haven’t been able to. Aliases can be assigned to accounts only on create. So they might want to create a new account id with an alias.

Reviewer 3: Perhaps just mention council members want to change their account id. alias account is just one specific case falls under the category

Author: I agree this is an example of a time the council members might want to change the account Id. Don't know all the use cases , so have added an actual case here. I feel this will give a better understanding of the rationale. Maybe I can add `As an example` there

- The idea of introducing rollups is to make it faster for new mirror nodes to download historical files starting from epoch. This will also help us, avoid having to rename millions of files. Renaming millions of files has to be done one by one via S3 APIs and would take months. With rollup we can create separate aggregated files while leaving individual files intact for longer period until we've verified authenticity of rollup.

## User stories

- As a mirror node operator, I'd like to be compatible with both the historical and new bucket path formats.
- As a node operator, I'd like to be able to change my node's account ID.
- As a network operator, I'd like to reduce cloud storage costs by eliminating duplicate files and compressing historical data.
- As a new mirror node operator I'd like to be able to download historical data in the new aggregated format. This will improve the time it takes to startup a new mirror node with all historical data synchronized and to improve download costs to bootstrap a new mirror node.

> 

Author: I am not sure what you mean.  Do you mean we should support sharding as part of this HIP?

> 

Reviewer 3: user story, the bucket struct needs to support sharding

> 

Author: The bucket structure does support sharding out of the box for eg.` {shard}/{nodeID}/balance/YYYY/MM/DD`. Not sure if I am missing something , but what do we need to explicitly do for it? Do you mean handle this in mirror_node, when reading from the node_id based path?

> 

Reviewer 3: all i'm saying is support sharding in the bucket structure is one of the user stories and thus we should add it here

   
In the existing scenario the default pathType would be `ACCOUNT_ID`, which would then be changed to AUTO when we are close to approaching Phase two.
After a significant amount of time has passed after launching Phase two, we will remove these properties to default the downloading from the node ID based path.

#### Phase Two:
Update the consensus nodes to write stream files to the new `{shard}/{nodeID}/*` nodeId based path.

#### Phase Three:
Create a new lambda function which performs data rollup along with compression.

##### Compression zstd vs gz:
- zstd has a better compression ration that gzip.
- (zstd:gzip)
  - Faster compression - About a 5:1  ratio.
  - Faster decompression - About a 3:1 ratio.

##### Algorithm:
- Read stream files from current location 
    ``` 
        accountBalances/balance0.0.3/ 
        eventsStreams/events_0.0.3/     
        recordstreams/record0.0.3/ 
    ```
    
- The new path for aggregated files would be:
    
    ``` 
        {shard}/record/
        {shard}/balance/
        {shard}/event/
    ```
- The lambda will decompress the file, if it is compressed.
- Calculate a hash of the file.

> - Maintain an in-memory map which has the hashes for the current file being processed.

Author: This is to check for duplication of files. Making sure we don't have duplicate files from the aggregated files.

> - Maintain an in-memory map which has the hashes for the current file being processed.

Reviewer: Non signature files

> - Maintain an in-memory map which has the hashes for the current file being processed.

Reviewer 3: problem is if we just look at the data file hashes and in case a group of, e.g., record files have two set of hashes, we may end up putting the one without super-majority. the uniqueness is already in the filename since it has the timestamp.

> - Maintain an in-memory map which has the hashes for the current file being processed.

Reviewer: Why is that a problem? That's how the data is now. It's the responsibility of the mirror node to download and verify. We don't want to discard non-majority files as that is a destructive operation where we lose data. We only want to discard duplicates based upon file hash. So if a specific filename resolves one file to hash A and 26 to hash B, we would upload one hash A and one hash B.

> - Maintain an in-memory map which has the hashes for the current file being processed.

Reviewer 3: not sure if I missed sth from the compressed file structure, what I get is for the same timestamp, there will be only one data file since data files are not put into node id specific path, only signature files are.

- Compress the file using zst if it not a signature file.
- Monitor the size/time of the potential aggregated file.
- For aggregation:
  - Calculate the node Id based on the current node account id. NodeId = (number portion of) nodeAccountId - 3
  ``` Node for balance0.0.4 would be 4-3 = 1  ```
  - The new compressed file structure would be as follows:
    Example:
    ```

>             2023-01-11T23_21.tar.zst

Author: Yes we are compressing individual files and then compress the tarball.

>             2023-01-11T23_21.tar.zst

Reviewer 3: why? generally speaking, compressing compressed data won't reduce size much? I assume we think the tarball can be further compressed, any data to support the idea?

>             2023-01-11T23_21.tar.zst

Author: Yes. I did actually compress these compressed files and though , the 2nd compression is not significant, there is saving there.
For a 24.5MB folder with several record files etc.
Individual files compressed with zstd and the folder tarred without zstd algorithm - 8.9 MB
Individual files compressed with zstd and the folder tarred with zstd algorithm - 8.2 MB

Like I said it is not a lot of difference , but there is a difference. Hence, I decided to go with this.

>             2023-01-11T23_21.tar.zst

Reviewer 3: have you considered the extra time needed for the second round compression and decompression? I don't see why it's worth

>             2023-01-11T23_21.tar.zst

Reviewer 3: alternatively, we can choose to leave all the files inside the tar ball decompressed, tar them, and compress the tar file. I do suspect in case we have many data files, the compression ratio will be greater the current design.

>             2023-01-11T23_21.tar.zst

Author: The time for further compression and decompression wasn't significant either and hence thought it is better to save some storage.But it is an extra step to decompress when we untar. The savings was 2-3%.

>             2023-01-11T23_21.tar.zst

Author: It will probably be devops who would do it right?

>             2023-01-11T23_21.tar.zst

Author: @xin-hedera I have edited the file accordingly to remove outside compression.

               - 2023-01-11T23_21_52.007485807Z.rcd.zst
               - 2023-01-11T23_21_50.006107818Z.rcd.zst
               - 2023-01-11T23_21_46.003350460Z.rcd.zst

>                - 2023-01-11T23_21_44.001332036Z.rcd.zst

Reviewer: Need to mention sidecar files

               - 0(This is the node id)
                 - 2023-01-11T23_21_52.007485807Z.rcd_sig
                 - 2023-01-11T23_21_50.006107818Z.rcd_sig
                 - 2023-01-11T23_21_46.003350460Z.rcd_sig
                 - 2023-01-11T23_21_44.001332036Z.rcd_sig
     ```
     The aggregation in the new files would be size/time based, as will be determined by performance testing.
- Once the limit has been reached, aggregate the file and write it to the new bucket location. 

> - Delete the existing files that have been read from the current bucket location.

Reviewer 3: there is a large piece missing in the rollup algorithm: how to pick the data file?

The straightforward way without the involvement of address book and file hash and signature verification is for each timestamp group, calculate the hashes of the data file, and pick the one with super majority.

Reviewer 3: the above comment should no longer apply since we are going to keep all data files

Reviewer: Not going to keep all, just all files with distinct contents.

## Motivation

Node account IDs are Hedera account IDs, which each node has obtained. These can be changed over time.
In the future we plan to introduce community nodes, thus increasing the number of unique nodes across the Hedera network. Granting specific node ID based access to buckets will be of utmost importance for security reasons.
The current folder structure uses an account ID per node. Account IDs are not ideal as a council member will want to change their account ID over time. To future-proof this, we will want to move from account IDs to shard and node ID in the path.

## Rationale

- There have been council members that want to change their account id but haven’t been able to. Aliases can be assigned to accounts only on create. So they might want to create a new account id with an alias.
- There might also be a need to recreate an account for security reasons.
- By adapting the new node ID based path we provide the council members the flexibility to change their account IDs as well as allow only node specific access to the buckets for future community nodes inclusion.

> - Adding the network name to the path helps:

Author: @steven-sheehy Could you see if this makes sense for network name addition.

Reviewer: Looks good.

    - Create bucket paths for ephemeral networks
    - Combine prepod environments
    - Allow not having to recreate buckets for resettable networks   
- The idea of introducing rollups is to make it faster for new mirror nodes to download historical files starting from epoch. This will also help us, avoid having to rename millions of files. Renaming millions of files has to be done one by one via S3 APIs and would take months. With rollup we can create separate aggregated files while leaving individual files intact for longer period until we've verified authenticity of rollup.

## User stories

- As a mirror node operator, I'd like to be compatible with both the historical and new bucket path formats.
- As a node operator, I'd like to be able to change my node's account ID.
- As a network operator, I'd like to create the new bucket structure that supports sharding.
- As a network operator, I'd like to reduce cloud storage costs by eliminating duplicate files and compressing historical data.

> - As a new mirror node operator I'd like to be able to download historical data in the new aggregated format. This will improve the time it takes to startup a new mirror node with all historical data synchronized and to improve download costs to bootstrap a new mirror node.

Reviewer: Could add a use case for network name.

## Specification

This HIP is to change the stream data bucket folder structure and in the process roll up data for faster ingestion.

Current Structure:
```
    accountBalances/balance{nodeAccountID}/
    eventsStreams/events_{nodeAccountID}/
    recordstreams/record{nodeAccountID}/
    
Example:
    accountBalances/balance0.0.3/
    eventsStreams/events_0.0.3/
    recordstreams/record0.0.3/
```
    
Proposed Structure:
```
    {network}/{shard}/{nodeID}/balance/YYYY/MM/DD
    {network}/{shard}/{nodeID}/event/YYYY/MM/DD
    {network}/{shard}/{nodeID}/record/YYYY/MM/DD
    

> Example:

Author: @steven-sheehy Could you see if this makes sense for network name addition.

Reviewer: Looks good

    Standard network:
    mainnet/0/3/balance/2020/05/24
    mainnet/0/3/event/2020/05/24
    mainnet/0/3/record/2020/05/24
    
    Resettable network:
    testnet-2023-01/0/3/balance/2020/05/24
    testnet-2023-01/0/3/event/2020/05/24
    testnet-2023-01/0/3/record/2020/05/24        
```
##### Problems with the current structure naming:
- Stream name is duplicated in `eventsStreams` and `recordstreams`. We are missing stream from `accountBalances`.
- Casing is inconsistent in `eventsStreams` and `recordstreams`.
- After the separator we are repeating stream names. `events_0.0.3` has an underscore where as the others don't.

> The new structure proposes consistent naming.In the future when we start supporting multiple shards, the new bucket structure will automatically work for it without any required changes.

Reviewer: nit
```suggestion
The new structure proposes consistent naming. In the future when we start supporting multiple shards, the new bucket structure will automatically work for it without any required changes.
```

> ##### Proposed naming structure:

Author: @steven-sheehy Could you see if this makes sense for network name addition.

Reviewer: Made a suggestion below.

- The network name in the bucket path will not change for non-resettable networks. 

> - For networks that can be reset the network key in the path will start wth the network name prefix followed by a timestamp, as stated in the above testnet example. On a reset, to get the latest network, one would need to sort the buckets and get the latest timestamp bucket to read from.

Reviewer: ```suggestion
- For networks that can be reset the network key in the path will start wth the network name prefix followed by a timestamp, as stated in the above testnet example. On a reset, to get the latest network one would need to filter by the network prefix (e.g. `testnet`) and sort the folders to get the latest network to read from.
```

- The granularity of the timestamp would depend on how often that particular network is reset.
- The remaining bucket path after the network name would be similar for all networks.

It is thus proposed to be a phased approach with three phases.

#### Phase One:
Update mirror node downloader to read from the new `{shard}/{nodeID}/*` nodeId based path as well as support reading from `*/*0.0.3/` account id based path.

##### Algorithm:
 - Add a `pathType` property with values `AUTO`, `ACCOUNT_ID`, and `NODE_ID` with a default of `ACCOUNT_ID`
 - Later once consensus nodes get closer to writing to the new path adjust default to `AUTO`
 - Add a `pathRefreshInterval` property with a default value of `1m`

>     If pathType is `AUTO`:

Reviewer: nit: Formatting error makes lines appear as one.
```suggestion
 - If pathType is `AUTO`:
```

    - Attempt to list signatures from  a path with node account IDs
    - If successful, update a per node timestamp and boolean to ensure account ID is used until pathRefreshInterval elapses
    - If no results, attempt to list signatures from a path with node IDs
    - If successful, set `pathType` as `NODE_ID` so the node ID path is used going forward
   
In the existing scenario the default pathType would be `ACCOUNT_ID`, which would then be changed to AUTO when we are close to approaching Phase two.

> After a significant amount of time has passed after launching Phase two, we will remove these properties to default the downloading from the node ID based path.

Reviewer: We can't remove this logic until phase three completes and all data is either using node ID or been aggregated.

Author: Moved it to after phase three

#### Phase Two:
Update the consensus nodes to write stream files to the new `{shard}/{nodeID}/*` nodeId based path.

#### Phase Three:
Create a new lambda function which performs data rollup along with compression.

##### Compression zstd vs gz:
- zstd has a better compression ration that gzip.
- (zstd:gzip)
  - Faster compression - About a 5:1  ratio.
  - Faster decompression - About a 3:1 ratio.

##### Algorithm:
- Read stream files from current location 
    ``` 
        accountBalances/balance0.0.3/ 
        eventsStreams/events_0.0.3/     
        recordstreams/record0.0.3/ 
    ```
    
- The new path for aggregated files would be:
    
    ``` 
        {shard}/record/
        {shard}/balance/

>         {shard}/event/

Reviewer: Need to update with network

- Compress the file using zst if it not a signature file.
- Monitor the size/time of the potential aggregated file.
- For aggregation:
  - Calculate the node Id based on the current node account id. NodeId = (number portion of) nodeAccountId - 3
  ``` Node for balance0.0.4 would be 4-3 = 1  ```
  - The new file structure would be as follows:
    Example:
    ```
    >tar -tf 2022-07-13T08_46.tar | sort  

    2022-07-13T08_46/
    2022-07-13T08_46/0/
    2022-07-13T08_46/0/2022-07-13T08_46_08.041986003Z.rcd.zst
    2022-07-13T08_46/0/2022-07-13T08_46_08.041986003Z.rcd_sig
    2022-07-13T08_46/0/2022-07-13T08_46_11.304284003Z.rcd.zst
    2022-07-13T08_46/0/2022-07-13T08_46_11.304284003Z.rcd_sig
    2022-07-13T08_46/0/2022-07-13T08_46_11.304284003Z_01.rcd.zst
    2022-07-13T08_46/1/
    2022-07-13T08_46/1/2022-07-13T08_46_08.041986003Z.rcd.zst
    2022-07-13T08_46/1/2022-07-13T08_46_08.041986003Z.rcd_sig
    
     ```
     The above compressed tar file includes the following files in node id based folders:
     - Record stream files (.rcd.zst)
     - Signature files (.rcd_sig)
     - Sidecar files (_01.rcd.zst)
     

>      NOTE:File 2022-07-13T08_46_08.041986003Z.rcd.zst appears in 2 separate node folders. These are files with distinct content and when downloaded by the mirror node one of the two file will be selected based on the consensus majority.

Reviewer: ```suggestion
     NOTE: File `2022-07-13T08_46_08.041986003Z.rcd.zst` appears in two separate node folders. These are files with distinct content and when downloaded by the mirror node one of the two file will be selected based on the consensus majority.
```

- As a node operator, I'd like to be able to change my node's account ID.
- As a network operator, I'd like to create the new bucket structure that supports sharding.
- As a network operator, I'd like to merge buckets for all prepod environments and be able to use bucket paths for ephemeral networks.
- As a network operator, I'd like to reduce cloud storage costs by eliminating duplicate files and compressing historical data.
- As a new mirror node operator I'd like to be able to download historical data in the new aggregated format. This will improve the time it takes to startup a new mirror node with all historical data synchronized and to improve download costs to bootstrap a new mirror node.

## Specification

This HIP is to change the stream data bucket folder structure and in the process roll up data for faster ingestion.

Current Structure:
```
    accountBalances/balance{nodeAccountID}/
    eventsStreams/events_{nodeAccountID}/
    recordstreams/record{nodeAccountID}/
    
Example:
    accountBalances/balance0.0.3/
    eventsStreams/events_0.0.3/
    recordstreams/record0.0.3/
```
    
Proposed Structure:
```

>     {network}/{shard}/{nodeID}/balance/

Reviewer 4: A shortcoming in our existing structure is that by virtue of being very flat, it is not possible to list the contents of these buckets - it would take too long.

Perhaps we should bucket these by year, month and day as well? That would facilitate these being explorable by humans. Today this is accomplished by knowing the format and making queries with wildcards. I think there is room to improve this experience.

Reviewer: We had that and removed it. It's not useful since we'll be rolling these up and deleting them soon after they're created. So there will be a bunch of empty folders.

Besides, it's not true that you can't list the contents of these buckets quickly. In S3 API you specify a marker or prefix to quickly filter out results after. That's how the mirror node efficiently searches for files after a certain date. So you can always list just the files for a certain year, month or day.

Author: @dalvizu Does the discussion we had yesterday and above answer your query?

Reviewer 4: It does, thanks

Author: Thanks Dan.

### Mirror Node Implications
The existing mirror nodes would need to implement phase one.

### Dev ops Implications
After the completion of phase one and having being notified, dev ops will need to implement phase two.

## Backwards Compatibility

This HIP proposes breaking changes. All existing mirror node operators would need to change code to support reading from the new as well as the old bucket location. According to the current plan for phase one and two, operators will have ample time until the completion of phase two and a significant amount of notification period, before the mirror nodes will become backwards incompatible.

## Security Implications

With this change, we are improving security. Previously, each node had to be granted access to three separate paths since we had stream types as the top level folders.
Write access to the new node id based path will be granting access to only specific node path buckets. This would prevent a malicious node to have write access or corrupt other stream files.

> The rollup process could be seen as centralized and a single point of failure. Since, the lambda will be deleting files it has already aggregated, this might be pursued as a problem.

Reviewer 4: Characterizing this as "seen as" implies it is a perception problem. The problem isn't perception though. A single cloud function in a single cloud provider is both centralized and a single point of failure and this is not a perception problem.

Have we discussed or considered how to implement a decentralized solution? Just throwing something out there: We have all of these consensus nodes that can certainly do the roll up and validate a roll up. Perhaps we pick some area where they can all write to, and allow them to delete from their own paths, and then a rollup process would be done on each node: to perform the process and dump results to their own path. Then check to see if everyone else has done this. And if they have and we're the last one, go ahead and write it to this central path and delete your own copy. And periodically check for these rollups if you've already performed your own, and if so validate it. There'd be some concurrency issues to work around, and we'd need to solve for bad actors, but this would mean no central actors and only nodes can delete their own data.

Reviewer: I might be misunderstanding you, but having each node do the rollup would be incredibly inefficient and negate the advantage of the rollup. The advantage of the rollup is that it dedupes the large stream files across nodes and that it groups the signatures files together. Sounds like you're suggesting only aggregating within a node. Which would be mostly useless for mirror nodes as we'd have to download at least 1/3 of the very large aggregated files for a given period before we could reach consensus on those files.

Reviewer 4: Based on discussion it sounds like the centralization is acceptable as by using a cloud provider we aren't making things any worse, and that the long term solution is for a council hosted file service
